File: UIViewControllerTransitions/Classes/AnimatedNavigationTransitioning.m
//
//  AnimatedNavigationTransitioning.m
//  UIViewControllerTransitions
//
//  Created by Steve Kim on 8/13/17.
//
//

#import "AnimatedNavigationTransitioning.h"

@implementation AnimatedNavigationTransitioning

#pragma mark - Overridden: AbstractAnimatedTransitioning

- (UIViewController *)aboveViewController {
    return _push ? self.toViewController : self.fromViewController;
}

- (UIViewController *)belowViewController {
    return _push ? self.fromViewController : self.toViewController;
}

- (void)animateTransition:(id <UIViewControllerContextTransitioning>)transitionContext {
    [super animateTransition:transitionContext];
    
    if (_push) {
        [self animateTransitionForPush:transitionContext];
    } else {
        [self animateTransitionForPop:transitionContext];
    }
}

#pragma mark - Protected methods

- (void)animateTransitionForPop:(id<UIViewControllerContextTransitioning>)transitionContext {
}

- (void)animateTransitionForPush:(id<UIViewControllerContextTransitioning>)transitionContext {
}

@end

File: AnimatedTransitionKit/Classes/Transitions/Base/AnimatedNavigationTransitioning.swift
//
//  AnimatedNavigationTransitioning.swift
//  Pods
//
//  Created by Minwoo on 9/4/25.
//

import UIKit

open class AnimatedNavigationTransitioning: AbstractAnimatedTransitioning {
    public var appearanceOption: TransitioningAnimationOptions?
    public var disappearanceOption: TransitioningAnimationOptions?
    
    open var isPush: Bool = false {
        didSet {
            options = isPush ? appearanceOption : disappearanceOption
        }
    }

    // MARK: - Overridden: AbstractAnimatedTransitioning

    override public var aboveViewController: UIViewController? {
        return isPush ? toViewController : fromViewController
    }

    override public var belowViewController: UIViewController? {
        return isPush ? fromViewController : toViewController
    }

    override public func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
        super.animateTransition(using: transitionContext)

        if isPush {
            animateTransition(forPush: transitionContext)
        } else {
            animateTransition(forPop: transitionContext)
        }
    }

    // MARK: - Protected methods

    open func animateTransition(forPop transitionContext: UIViewControllerContextTransitioning) {
        // To be implemented by subclasses
    }

    open func animateTransition(forPush transitionContext: UIViewControllerContextTransitioning) {
        // To be implemented by subclasses
    }
}
